### DISCUSSION TITLE
Does the web socket transport support ping/pong frames?

### DISCUSSION THREAD

**Discussion by 0xGuybrush** [wevm/viem#1591] [Question]
I'm trying to understand how/if the web-socket transport layer supports [Ping/Pong control frames](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets).

I had a read through the code in [`rpc.ts`](https://github.com/wevm/viem/blob/main/src/utils/rpc.ts#L183-L191) (and took a look too at [isows repo](https://github.com/wevm/isows)) but couldn't see where this was being handled.

If it is being handled, could someone point me in the right direction of the file so I can understand it? 🙏 

We're working against a private chain and have noticed connections getting dropped — I thought at first it was a connectivity issue (like issue https://github.com/wevm/viem/issues/877 mentions) but if I added in a call for `getLatestBlock()` every N seconds and haven't had any issues since. 

Which got me wondering if the cause is that we're not responding to `PING`s & the server is closing the connection and the reason we don't see the issue on public chains is because the baseline level of activity is naturally keeping the connection alive?

**Comment by nstylo** [comment]
I am wondering the same. I am using `watchContractEvent` using a websocket client. My node provider closes a websocket connection after one hour, so I have to make sure to keep it alive. It would be great if some sort of client side keep alive mechanism was put in place.
